src/extension/gogoduck/src/main/java/au/org/emii/gogoduck/worker/GoGoDuck.java

```diff
@@ -0,0 +1,511 @@
+                }
+            }
+
+            String extension = FilenameUtils.getExtension(dst.getFileName().toString());
+
+            if (extension.equals("gz")) {
+                gunzip(dst.toFile());
+            }
+        }
+    }
+
+    private static URL fileURItoURL(URI uri) {
+        try {
+            String uriStr = uri.toString();
+            for (String key : replacePrefixes.keySet()) {
+                if (uriStr.startsWith(key)) {
+                    uriStr = uriStr.replace(key, replacePrefixes.get(key));
+                }
+            }
+            return new URL(uriStr);
+        }
+        catch (MalformedURLException e) {
+            throw new GoGoDuckException(e.getMessage());
+        }
+    }
+
+    private static void gunzip(File file) {
+        try {
+            logger.info(String.format("Gunzipping '%s'", file));
+            File gunzipped = File.createTempFile("tmp", ".nc");
+
+            FileInputStream fis = new FileInputStream(file);
+            GZIPInputStream gis = new GZIPInputStream(fis);
+            FileOutputStream fos = new FileOutputStream(gunzipped);
+            byte[] buffer = new byte[8192];
+            int len;
+            while((len = gis.read(buffer)) != -1){
+                fos.write(buffer, 0, len);
+            }
+            fos.close();
+            gis.close();
+
+            Files.delete(file.toPath());
+            Files.move(gunzipped.toPath(), file.toPath());
+        } catch (IOException e) {
+            throw new GoGoDuckException(String.format("Failed gunzip on '%s': '%s'", file, e.getMessage()));
+        }
+    }
+
+    private static void applySubsetSingleFileNcks(File file, GoGoDuckModule module) {
+        List<String> ncksSubsetParameters = module.getSubsetParameters().getNcksParameters();
+        List<String> ncksExtraParameters = module.ncksExtraParameters();
+
+        try {
+            File tmpFile = File.createTempFile("tmp", ".nc");
+
+            List<String> command = new ArrayList<String>();
+            command.add(ncksPath);
+            command.add("-a");
+            command.add("-4");
+            command.add("-O");
+            command.addAll(ncksSubsetParameters);
+            command.addAll(ncksExtraParameters);
+
+            command.add(file.getPath());
+            command.add(tmpFile.getPath());
+
+            logger.info(String.format("Applying subset '%s' to '%s'", ncksSubsetParameters, file.toPath()));
+            execute(command);
+
+            Files.delete(file.toPath());
+            Files.move(tmpFile.toPath(), file.toPath());
+        }
+        catch (Exception e) {
+            throw new GoGoDuckException(String.format("Could not apply subset to file '%s': '%s'", file.getPath(), e.getMessage()));
+        }
+    }
+
+    private static void applySubsetSingleFileNative(File file, GoGoDuckModule module) {
+        // TODO implement!
+    }
+
+    private static class ncksRunnable implements Runnable {
```

Comment:
As mentioned above, I definitely think that this code to do with doing the actual subsetting could live in its own class (or classes, possibly one for the ncks specific stuff too).